File: packages/js-sdk/src/types/quote-line-item.ts
/**
 * TypeScript types for TurboQuote — Line Item entity and request/response types
 */

import type {
  BillingFrequency,
  DiscountType,
  LineItemType,
  PaginatedResponse,
} from './quote-shared';
import type { Product } from './product';

// ============================================
// DOMAIN TYPES
// ============================================

export interface LineItem {
  id: string;
  orgId: string;
  quoteId: string;
  lineItemType: LineItemType;
  parentLineItemId: string | null;
  productId: string | null;
  productName: string | null;
  productSku: string | null;
  productDescription: string | null;
  bundleId: string | null;
  bundleName: string | null;
  bundleDescription: string | null;
  quantity: number;
  unitPrice: number;
  discountPercent: number;
  discountType: DiscountType | null;
  discountAmount: number | null;
  displayOrder: number | null;
  subtotal: number;
  cost: number | null;
  marginPercent: number | null;
  categoryId: string | null;
  categoryName: string | null;
  billingFrequency: BillingFrequency | null;
  showItemsToEndUser: boolean;
  isActive: boolean;
  createdBy: string | null;
  createdOn: string;
  updatedOn: string;
  product?: Product;
  childLineItems?: LineItem[];
}

// ============================================
// REQUEST TYPES
// ============================================

export interface AddLineItemRequest {
  productId: string | null;
  productName: string;
  unitPrice: number;
  billingFrequency: BillingFrequency;
  quantity?: number;
  discountPercent?: number;
  discountType?: DiscountType;
  discountAmount?: number;
  categoryId?: string | null;
  categoryName?: string | null;
  cost?: number | null;
  productSku?: string | null;
  productDescription?: string | null;
}

export interface AddBundleLineItemRequest {
  bundleId: string;
  bundleName: string;
  quantity?: number;
  discountPercent?: number;
  discountType?: DiscountType;
  discountAmount?: number;
  bundleDescription?: string | null;
  showItemsToEndUser?: boolean;
}

export interface UpdateLineItemRequest {
  quantity?: number;
  unitPrice?: number;
  discountPercent?: number;
  discountType?: DiscountType;
  discountAmount?: number;
  displayOrder?: number | null;
  billingFrequency?: BillingFrequency;
  categoryId?: string | null;
  categoryName?: string | null;
  cost?: number | null;
  showItemsToEndUser?: boolean;
  productName?: string;
  productSku?: string | null;
  productDescription?: string | null;
}

export interface ListLineItemsOptions {
  limit?: number;
  offset?: number;
  lineItemType?: LineItemType;
  billingFrequency?: BillingFrequency;
  parentLineItemId?: string;
}

// ============================================
// RESPONSE TYPES
// ============================================

export type LineItemListResponse = PaginatedResponse<LineItem>;

File: packages/py-sdk/src/turbodocx_sdk/types/quote_line_item.py
"""
TypedDict definitions for TurboQuote -- Line Item entity.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional
from typing_extensions import TypedDict

from .product import Product


class LineItem(TypedDict, total=False):
    id: str
    orgId: str
    quoteId: str
    lineItemType: str  # LineItemType literal
    parentLineItemId: Optional[str]
    productId: Optional[str]
    productName: Optional[str]
    productSku: Optional[str]
    productDescription: Optional[str]
    bundleId: Optional[str]
    bundleName: Optional[str]
    bundleDescription: Optional[str]
    quantity: int
    unitPrice: float
    discountPercent: float
    subtotal: float
    cost: Optional[float]
    marginPercent: Optional[float]
    categoryId: Optional[str]
    categoryName: Optional[str]
    billingFrequency: Optional[str]  # BillingFrequency literal
    showItemsToEndUser: bool
    isActive: bool
    createdBy: Optional[str]
    createdOn: str
    updatedOn: str
    product: Product
    childLineItems: List[LineItem]


class _AddLineItemRequestRequired(TypedDict):
    productName: str
    unitPrice: float
    billingFrequency: str  # BillingFrequency literal


class AddLineItemRequest(_AddLineItemRequestRequired, total=False):
    productId: Optional[str]
    quantity: int
    discountPercent: float
    categoryId: Optional[str]
    categoryName: Optional[str]
    cost: Optional[float]
    productSku: Optional[str]
    productDescription: Optional[str]


class _AddBundleLineItemRequestRequired(TypedDict):
    bundleId: str
    bundleName: str


class AddBundleLineItemRequest(_AddBundleLineItemRequestRequired, total=False):
    quantity: int
    discountPercent: float
    bundleDescription: Optional[str]
    showItemsToEndUser: bool


class UpdateLineItemRequest(TypedDict, total=False):
    quantity: int
    unitPrice: float
    discountPercent: float
    billingFrequency: str  # BillingFrequency literal
    categoryId: Optional[str]
    categoryName: Optional[str]
    cost: Optional[float]
    showItemsToEndUser: bool
    productName: str
    productSku: Optional[str]
    productDescription: Optional[str]


class ListLineItemsOptions(TypedDict, total=False):
    limit: int
    offset: int
    lineItemType: str  # LineItemType literal
    billingFrequency: str  # BillingFrequency literal
    parentLineItemId: str


class LineItemListResponse(TypedDict):
    results: List[LineItem]
    totalRecords: int
